fix!: disable transaction commits during doc events

ankush opened the pull request:
- Events like doc.save and doc.submit need to be atomic
- Document hooks can make it not so atomic.

This is extending server script behavior where server script hooks are
not allowed to commit/rollback.

Summary:
- First party code on controller CAN do transaction commits
- Third party code or ANY events triggered via hooks can not do transaction commits.  



Example problem:

![image](https://github.com/frappe/frappe/assets/9079960/b79aa51a-3c0e-4f2b-8023-71235758ae6f)

ankush committed:
fix: disable transaction commits during doc events

- Events like doc.save and doc.submit need to be atomic
- Document hooks can make it not so atomic.

This is extending server script behaviour where server script hooks are
not allowed to commit/rollback.

cogk commented:
Hey, I'm not sure that I completely understand the implications of this change (which looks good to me).  
Yet, I'm wondering if this won't cause issues when updating the `status` of a document using a `db_set`… let me explain.  
Let's imagine a "Online Payment" document `doc` that has a **hookable** method `on_payment_done`.

```python
class OnlinePayment(Document):
    status: Literal["Pending", "Paid"]

    def on_payment_done(self):
        if doc.status == "Paid": return
        doc.db_set("status", "Paid", commit=True)  # I really want to write down that the customer has paid
        self.do_risky_stuff()  # Here, if an exception is thrown, won't this prevent the previous change being committed?

@frappe.whitelist()
def stripe_webhook(data):
    doc = frappe.get_doc("Online Payment", ...)
    doc.run_method("on_payment_done")  # <-- here we disable_transaction_control

# When the customer pays online, Stripe calls the `app.….tripe_webhook` method
# which triggers the `on_payment_done` method via `run_method`.
```

ankush commented:
Yep, this has it's downsides. I am not sure how to restrict it correctly.

Maybe I should do this for very specific events? Those events would be between doc updated and db and on_submit completion usually. Anything before that can be fine and anything after that can be fine usually.

ankush commented:
The problem is idiots writing db.commit from on_submit and then if exception occurs the doc is submitted without doing on submit actions.


Usually, in most cases savepoints can be better alternative for things that can potentially fail but need to still let control flow progress. There's no way developers can ensure atomicity if they use commit/rollback otherwise.

ankush commented:
![9k=(10)](https://github.com/frappe/frappe/assets/9079960/f53b41ba-05d3-4789-85cf-ead39b4ecfaa)

ankush committed:
fix: handle nested event calls

Treat disable_transaction_control as a stack incr/decr when moving in
and out of context.

ankush commented:
Latest iteration of problems caused by db.commit

![image](https://github.com/frappe/frappe/assets/9079960/9e5549f4-1ac6-470b-b31a-f8973309fa45)

ankush commented:
@cogk we'll really have to disable commits completely. There's just no sane way to ensure atomicity and preventing concurrent usage without it.

The alternate for pattern you shown is -> do minimal action in event and enqueue risky processing in background.

ankush commented:
> The alternate for pattern you shown is -> do minimal action in event and enqueue risky processing in background.

You HAVE to do this anyway in some cases. Payment gateways require you to respond in some time so those actions shouldn't have lot of work built into it, risky or not. Alternatively, don't do `run_method` and call the method directly.

ankush committed:
test: transaction control
